### PR TITLE
activity node missing  isForCompensation property And association "attachedToRef " attributes are also missing

### PR DESCRIPTION
**Pull request by qiudaoke**

### CODE REVIEW

**Comment by wberges** [comment]
Hello,
I confirm that Compensation isn't well managed (6.4.0 version).
I have created a simple compensation workflow, checked it with the Check button of the editor, everything is ok.
But when I run my process, I have an error saying `Compensation activity could not be found (or it is missing 'isForCompensation="true"`'
I have tried to import the BPMN file in another BPMN editor, and some links are missing, in addition to the isForCompensation attribute.
 
simple_compensation_bpmn

And when I create a more complex workflow (several tasks with a compensation event for each one), there are a lot of links which are lost (`attachedToRef` attribute missing). In addition to isForCompensation... :)

Here is my simple BPMN file for the compensation: 
[simple_compensation.bpmn20.zip](https://github.com/flowable/flowable-engine/files/2530918/simple_compensation.bpmn20.zip)

HTH

**Comment by qiudaoke** [comment]
hi,
@wberges  

I imported this process into the modeler designer. On this PR branch, I added an Is for compensation attribute. Then I exported the process, he can run normally, and the following attribute isForCompensation="true" in the xml.

@tijsrademakers 
Can you merge this PR? Thank you

**Comment by wberges** [comment]
Hello,
I have designed a little bit more complex compensation diagram.
image
This time, it shows that some "attachedToRef " attributes are also missing.
[Compensation.zip](https://github.com/flowable/flowable-engine/files/2543665/Compensation.zip)
@qiudaoke 
Perhaps good to test your PR using this diagram before the merge ? :)
Cheers

**Comment by qiudaoke** [comment]
@wberges 
I imported this model（Compensation.zip）, but he reported it wrong。
java.lang.NullPointerException: null
	at org.flowable.ui.modeler.service.BpmnDisplayJsonConverter.fillWaypoints(BpmnDisplayJsonConverter.java:311) ~[classes/:?]

Can you re-upload the correct model?

**Comment by wberges** [comment]
Hi @qiudaoke 
Unfortunately, it is the correct model... :(
I just tested again (without App, just the bpmn):
- opened my process
image

- checked it (no error, green)
- exported it
- re-imported it, and the result is the following one:

image
The process is completely destroyed.
I think it's a bug from the modeler, certainly linked to the missing attributes.
If I open it in another BPMN modeler, here is what is loaded:
image
No more COMP_3 due to the attribute `attachedToRef="SCRIPT_3"` missing (why only the third one???)
image
And COMP_SCRIPT_XX are no more Compensation tasks (attributes  `isForCompensation="true"` missing)
So, to summarize, as I said, it's not only `isForCompensation` attributes which are missing... :(
I will try later to remodel manually my BPMN and post it.

**Comment by wberges** [comment]
Hi @qiudaoke
Sorry, I have tried manually and also using another BPMN editor, but no way to import it correctly.
There are always some components which are badly imported (cancel event no more attached to the transaction sub-process, error event missing...).
After several tests, I think that the best and fastest way is to do a new diagram in your own editor and try to export and reimport it.

**Comment by wberges** [comment]
After additional investigations, it seems that graphical parts of the links aren't well defined/exported by the modeler.
I think that links are supposed to be Edges, when Tasks are Shapes (I'm not a BPMN language expert :)).
But after exporting a simple model, I have both for the **compensation** link:
image

```
      <bpmndi:BPMNShape id="**BPMNShape_COMP1_TO_COMPTASK**" bpmnElement="COMP1_TO_COMPTASK">
        <omgdc:Bounds x="416.6299428734836" y="234.89337256954119" width="0" height="34.64393172577647" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNEdge id="**BPMNEdge_COMP1_TO_COMPTASK**" bpmnElement="COMP1_TO_COMPTASK">
        <omgdi:waypoint x="416.6299428734836" y="234.74955042983564" />
        <omgdi:waypoint x="416.6299428734836" y="270" />
      </bpmndi:BPMNEdge>
```
It's why I was unable to import the BPMN file in another editor without errors.
I have deleted the `<bpmndi:BPMNShape>...<bpmndi:BPMNShape/>` section from the XML file and it is well imported now.
Another bug for compensation ? :)

**Comment by wberges** [comment]
After export from flowable modeler, I have done the following modifications:
- added attribute `attacheToRef="script_3"` to `<boundaryEvent id="COMP_3"` (again, why only the third one??)
- added `isForCompensation="true"` to the 3 compensation tasks (`<scriptTask id="COMP_SCRIPT_X1"`)
- deleted the 3 `<bpmndi:BPMNShape bpmnElement="COMPX_TO_COMPSCRIPTX1"`
- Another point: `isExpanded="true"` attribute isn't present in my (designed expanded) subprocess shape, it appears so not expanded in the other modeler. Added.
`<bpmndi:BPMNShape id="BPMNShape_SUB_PROCESS" bpmnElement="SUB_PROCESS" isExpanded="true">`

Considering that I'm unable to import them (pb with isForCompensation again), I post blind 2 BPMN (simple bpmn in a zip, not App): 
- Without Compensation link shapes:
[Complex_compensation.manualmodif.importInFlowable_NOSHAPE.bpmn20.xml.zip](https://github.com/flowable/flowable-engine/files/2550225/Complex_compensation.manualmodif.importInFlowable_NOSHAPE.bpmn20.xml.zip)
- With Compensation link shapes (in case of):
[Complex_compensation.manualmodif.importInFlowable_WITHSHAPES.bpmn20.xml.zip](https://github.com/flowable/flowable-engine/files/2550228/Complex_compensation.manualmodif.importInFlowable_WITHSHAPES.bpmn20.xml.zip)

@qiudaoke Could you try to import them in your PR please?
Thanks

**Comment by wberges** [comment]
Ok, I was able to import both BPMN (with and without Shapes) via the REST API and to execute them succesfully!
But still no way to import and see them in the modeler (waiting for the PR) :(
BTW, the diagram history in the engine isn't really clean, but it works...
image

**Comment by qiudaoke** [comment]
@wberges 
Yes, I found this to be a bug. So I reworked the implementation logic of the convertToJson method in the BaseBpmnJsonConverter class. I have modified some code on this pr.

**Comment by qiudaoke** [comment]
@tijsrademakers 
Can you merge this pr? Thank you

**Comment by tijsrademakers** [comment]
Thanks for the PR
